### DISCUSSION TITLE
Question: How to programmatically add item to BagPart

### DISCUSSION THREAD

**Discussion by LMSSonos** [OrchardCMS/OrchardCore#6927] [Q&A] [answered]
I have a ContentType ActivitiesPage which has a BagPart in it. In the BagPart I have multiple entries of the type ActivityDateItem. I have created a ActivityDateItem.cs model
```
public class ActivityDateItem : ContentPart
{
        public DateField ActivityDate { get; set; }
        public TextField Description { get; set; }
}
```

I load the ActivitiesPage through the following call
` ContentItem contentItem= _orchardHelper.GetContentItemByIdAsync("MyId1234", false).Result;`

After this, I try to add an item as following
```
// get bag part
var bagPart = contentItem.Get<BagPart>("BagPart");

// create new entry
var newItem = new ActivityDateItem();
newItem.Description = new TextField { Text = "test description" };
newItem.ActivityDate = new DateField { Value = DateTime.Now };

// create new content item
ContentItem ci = new ContentItem { ContentType = "ActivityDateItem" };

// weld the newItem to the contentItem
ci.Weld(newItem);

// weld the content item to the bagPart
bagPart.Weld<ContentItem>(ci);

// weld the bagpart to the main contentitem (the activitypage)
contentItem.Weld<BagPart>(bagPart);

// update and publish page
await _contentManager.UpdateAsync(contentItem);
await _contentManager.PublishAsync(contentItem);
```

This all goes through without any exception. But when calling again `GetContentItemByIdAsync` I still get the original version and the new ActivityDateItem is not saved. 

Thanks for your help

**Comment by Piedone** [comment]
This looks like something for [Discussions](https://github.com/OrchardCMS/OrchardCore/discussions) so I'll convert it into a discussion.

**Comment by deanmarcussen** [comment]
There's a few things here which don't seem right.

- Never create a new ContentItem directly, always use the content manager
- I don't see anywhere where you are updating the `BagPart.ContentItems` list
- There's a lot of `Welding` going on here which doesn't seem the way to go

Something like this makes more sense.

```
var activity = await _contentManager.NewAsync("ActivityDateItem");
activity.Alter<ActivityDateItem>(x => 
{
    x.Alter<TextField>("Description", d => d.Text = "SomeValue");
    ...
};

contentItem.Alter<BagPart>(x =>
{
   x.ContentItems.Add(activity);
}

await _contentManager.UpdateAsync(contentItem);
await _contentManager.PublishAsync(contentItem);

```

**Comment by LMSSonos** [comment]
thank you @deanmarcussen - this input was very helpful. So I can now add a new item into the list, this one works. What I saw when using NewAsync method I get the ContentType, but it contains (when checking the json in the debugger) only the contained parts, but not the fields. It's the same behavior when using ContentTypes from the built-in examples. 

So what happnes is in the second line of your code when casting to ActivityDateItem I get the following error: 
`InvalidCastException: Unable to cast object of type 'OrchardCore.ContentManagement.ContentPart' to type 'MySolution.Module.ContentTypes.ActivityDateItem'.`

As shown in my sample code, ActivityDateItem extends ContentPart - not sure if this is the issue. Thanks for helping

**Comment by deanmarcussen** [comment, marked as answer]
Yes, the fields aren't welded on till later, and it won't become an `ActivityDateItem` (Part) until it gets serialized there and back again, probably.

Try with
```
activity.Alter<ContentPart>("ActivityDateItem", x => 
{
    x.Alter<TextField>("Description", d => d.Text = "SomeValue");
    ...
};
```

While hard type parts seem like a useful thing, they were never designed to be that, (they work because of a side effect), so there are a bunch of things that don't necessarily behave how you would expect, and places where you want to use them, but you can't, because they haven't become that yet.

As much as I like my hard typed classes, I've tried them, and gone away from them for the most part, as some of the ways their behaviour changes, in particular with display management (they are no longer placed in a dynamic part shape, which I find useful)

**Comment by LMSSonos** [comment]
Thanks for the details. The data is now saved, but with a little issue that the ContentPart is dded next to the ActivityItem. From the Debugger

```
{{
  "ActivityDateItem": {},
  "ContentPart": {
    "DateDescription": {
      "Text": "test"
    },
    "DateTo": {
      "Value": "2020-08-20T00:00:00+02:00"
    },
    "DateFrom": {
      "Value": "2020-08-20T00:00:00+02:00"
    }
  }
}}
```

The item which has been generated within Orchard looks like this.

```
{{
  "ActivityDateItem": {
    "DateFrom": {
      "Value": "2020-11-28T00:00:00Z"
    },
    "DateDescription": {
      "Text": "test"
    },
    "DateTo": {
      "Value": "2021-04-08T00:00:00Z"
    }
  }
}}
```

Just to be sure. You are saying that Typed Classed are not supported (That's ok for me), but it is allowed to use the APIs to add new entries programmatically, right?

**Reply by deanmarcussen** [reply]
No that doesn't look right. I think you've missed the Name property on the `.Alter` extension

```
activity.Alter<ContentPart>("ActivityDateItem", <<<----- Is this bit missing?
x => 
{
    x.Alter<TextField>("Description", d => d.Text = "SomeValue");
    ...
};
```

Re: Typed classes. Just sometimes changes behaviour, and for me, have turned out to be less helpful than I thought they would be.
i.e. like your situation, where you have one, but can't really make use of in the code where you're at ;)

API's yes absolutely

**Reply by ns8482e** [reply]
Just a sidebar - @LMSSonos you can create an extension method that do `activity.Alter<ActivityDateItem>(x ==> { ... })` or fallback to 
  `activity.Alter<ContentPart>("ActivityDateItem", { ... })` and your main code reuse extension method for your `TypedContentTypes`

**Comment by LMSSonos** [comment]
Thanks very much, now all works as expected. Is this something you would like I make a entry into the how-to guides? https://docs.orchardcore.net/en/dev/docs/guides/
